File: src/comms/fs.ts
import * as nodeFs from "fs";
import * as nodePath from "path";

// #EXT-006-REQ-2 Start
/**
 * Shared file system — the second (and only other) sanctioned inter-agent
 * channel alongside the rigid queue (`./queue`). It is the durable exchange
 * surface: agents communicate by writing/reading artifacts under a fixed,
 * canonical layout, never by addressing one another directly.
 *
 * Canonical layout (workspace-relative, rooted at a configurable base dir):
 *
 *   <base>/state       durable agent/system state
 *   <base>/inbox       messages/work delivered to an agent
 *   <base>/outbox      messages/work produced by an agent
 *   <base>/artifacts   produced files (outputs, logs, blobs)
 *
 * Access rules ([REQ-2], [REQ-4]):
 *   - All exchange uses workspace-relative paths that resolve *within* the
 *     layout. Absolute paths and `..` traversal that escape the base are
 *     refused with a typed {@link LayoutViolationError} before any I/O.
 *   - The on-disk structure is validated against the canonical layout via
 *     {@link validateLayout}.
 *
 * The default base dir is `.jaros-data/` (workspace-relative); tests pass a
 * temp dir so they never touch the real workspace.
 */

/** The four canonical, workspace-relative top-level directories of the layout. */
export const LAYOUT_DIRS = ["state", "inbox", "outbox", "artifacts"] as const;

/** A canonical top-level layout directory name. */
export type LayoutDir = (typeof LAYOUT_DIRS)[number];

/** Default workspace-relative base directory for the shared file system. */
export const DEFAULT_BASE_DIR = ".jaros-data";

/**
 * Typed error raised on any access that violates the canonical layout: a write
 * outside the layout, a path that escapes the base via traversal/absolute
 * escape, or an on-disk structure that does not match. Layout violations fail
 * loudly ([REQ-4]).
 */
export class LayoutViolationError extends Error {
  /** The offending path or detail (kept for diagnostics). */
  readonly detail: string;

  constructor(message: string, detail: string) {
    super(message);
    this.name = "LayoutViolationError";
    this.detail = detail;
    Object.setPrototypeOf(this, LayoutViolationError.prototype);
  }
}

/**
 * The shared file system rooted at a single base directory. Every method
 * operates on workspace-relative paths and refuses anything that would resolve
 * outside the base.
 */
export class SharedFileSystem {
  /** Absolute, normalized base directory containing the canonical layout. */
  readonly baseDir: string;

  constructor(baseDir: string = DEFAULT_BASE_DIR) {
    this.baseDir = nodePath.resolve(baseDir);
  }

  /**
   * Create the base dir and all canonical layout directories if absent. Safe to
   * call repeatedly (idempotent). Returns this instance for chaining.
   */
  ensureLayout(): this {
    nodeFs.mkdirSync(this.baseDir, { recursive: true });
    for (const dir of LAYOUT_DIRS) {
      nodeFs.mkdirSync(nodePath.join(this.baseDir, dir), { recursive: true });
    }
    return this;
  }

  /**
   * Resolve a workspace-relative path to an absolute path that is guaranteed to
   * live within {@link baseDir}. Throws {@link LayoutViolationError} for
   * absolute inputs or any `..` traversal that escapes the base.
   */
  resolve(relativePath: string): string {
    if (typeof relativePath !== "string" || relativePath.length === 0) {
      throw new LayoutViolationError(
        "Path must be a non-empty workspace-relative string.",
        String(relativePath)
      );
    }
    if (nodePath.isAbsolute(relativePath)) {
      throw new LayoutViolationError(
        "Absolute paths are not permitted; use a workspace-relative path within the layout.",
        relativePath
      );
    }

    const resolved = nodePath.resolve(this.baseDir, relativePath);
    // The resolved path must be the base dir itself or strictly contained in it.
    const baseWithSep = this.baseDir.endsWith(nodePath.sep)
      ? this.baseDir
      : this.baseDir + nodePath.sep;
    if (resolved !== this.baseDir && !resolved.startsWith(baseWithSep)) {
      throw new LayoutViolationError(
        "Path escapes the shared file system layout.",
        relativePath
      );
    }
    return resolved;
  }

  /**
   * Read a UTF-8 file at a workspace-relative path within the layout. Throws
   * {@link LayoutViolationError} if the path escapes the layout.
   */
  read(relativePath: string): string {
    const abs = this.resolve(relativePath);
    return nodeFs.readFileSync(abs, "utf8");
  }

  /**
   * Write `data` to a workspace-relative path within the layout, creating
   * parent directories as needed. Refuses (throws {@link LayoutViolationError})
   * any path that resolves outside the layout — traversal or absolute escape.
   */
  write(relativePath: string, data: string): void {
    const abs = this.resolve(relativePath);
    nodeFs.mkdirSync(nodePath.dirname(abs), { recursive: true });
    nodeFs.writeFileSync(abs, data, "utf8");
  }

  /**
   * Assert the on-disk structure matches the canonical layout: the base dir and
   * each of the four canonical directories must exist as directories. Throws
   * {@link LayoutViolationError} describing the first missing/invalid entry.
   */
  validateLayout(): void {
    if (!isDirectory(this.baseDir)) {
      throw new LayoutViolationError(
        "Shared file system base directory is missing or not a directory.",
        this.baseDir
      );
    }
    for (const dir of LAYOUT_DIRS) {
      const full = nodePath.join(this.baseDir, dir);
      if (!isDirectory(full)) {
        throw new LayoutViolationError(
          `Canonical layout directory "/${dir}" is missing or not a directory.`,
          full
        );
      }
    }
  }
}

/** True when `p` exists and is a directory. */
function isDirectory(p: string): boolean {
  try {
    return nodeFs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}
// #EXT-006-REQ-2 End

File: jaros/comms/fs.py
"""Shared file system with a fixed, validated layout (EXT-006 / REQ-2, REQ-4, REQ-6).

The shared file system is the durable exchange surface between agents. It has a
canonical layout of top-level directories; all file exchange uses
workspace-relative paths *within* that layout. Writes that try to escape the
workspace (via ``..`` traversal or absolute paths) are refused loudly with a
typed :class:`LayoutViolationError`.

It is backed by local/mounted files only — **no database or external service**
(EXT-006 / REQ-6) — keeping the zero-infrastructure tenet intact.

Canonical layout (relative to ``base_dir``)::

    state/        durable state-machine data
    inbox/        messages awaiting processing
    outbox/       messages produced for others
    artifacts/    durable work products
    agents/      agent payloads / config
    processed/    successfully handled inputs
    failed/       inputs that failed handling
"""

from __future__ import annotations

import os
from pathlib import Path

# #EXT-006-REQ-2 Start
#: The canonical top-level directories of the shared file system layout.
LAYOUT_DIRS: tuple[str, ...] = (
    "state",
    "inbox",
    "outbox",
    "artifacts",
    "agents",
    "processed",
    "failed",
)


# #EXT-006-REQ-4 Start
class LayoutViolationError(ValueError):
    """Raised when a path escapes the workspace or the layout is invalid.

    Covers both ``..`` traversal / absolute-path escapes at access time and
    structural validation failures from :meth:`SharedFileSystem.validate_layout`.
    """
# #EXT-006-REQ-4 End


class SharedFileSystem:
    """Access API for the shared file system rooted at ``base_dir``.

    All reads and writes go through :meth:`read` / :meth:`write`, which resolve
    workspace-relative paths and refuse any path that resolves outside
    ``base_dir``. Writes are atomic (temp file + ``os.replace``).
    """

    #: Re-exported so callers can reference the layout without importing the module constant.
    LAYOUT_DIRS: tuple[str, ...] = LAYOUT_DIRS

    def __init__(self, base_dir: str | os.PathLike[str]) -> None:
        self.base_dir: Path = Path(base_dir).resolve()

    def ensure_layout(self) -> None:
        """Create ``base_dir`` and every canonical layout directory.

        Idempotent: existing directories are left untouched.
        """
        self.base_dir.mkdir(parents=True, exist_ok=True)
        for name in LAYOUT_DIRS:
            (self.base_dir / name).mkdir(parents=True, exist_ok=True)

    def _resolve(self, path: str | os.PathLike[str]) -> Path:
        """Resolve a workspace-relative ``path`` to an absolute path in base_dir.

        Raises:
            LayoutViolationError: if ``path`` is absolute or resolves (after
                ``..`` normalisation) outside ``base_dir``.
        """
        raw = Path(path)
        if raw.is_absolute():
            raise LayoutViolationError(
                f"absolute paths are not allowed: {path!r}"
            )
        candidate = (self.base_dir / raw).resolve()
        # Confine the resolved path strictly within base_dir (refuses `..` escapes).
        if candidate != self.base_dir and self.base_dir not in candidate.parents:
            raise LayoutViolationError(
                f"path escapes the workspace layout: {path!r}"
            )
        return candidate

    def read(self, path: str | os.PathLike[str]) -> str:
        """Read and return the text content of a workspace-relative ``path``.

        Raises:
            LayoutViolationError: if ``path`` escapes the workspace.
            FileNotFoundError: if the resolved file does not exist.
        """
        target = self._resolve(path)
        return target.read_text(encoding="utf-8")

    def write(self, path: str | os.PathLike[str], data: str) -> None:
        """Atomically write ``data`` to a workspace-relative ``path``.

        Parent directories within the workspace are created as needed. The write
        is atomic: data is written to a temp file in the same directory and then
        ``os.replace``-d into place.

        Raises:
            LayoutViolationError: if ``path`` escapes the workspace.
        """
        target = self._resolve(path)
        target.parent.mkdir(parents=True, exist_ok=True)
        tmp = target.with_name(f".{target.name}.{os.getpid()}.tmp")
        tmp.write_text(data, encoding="utf-8")
        os.replace(tmp, target)

    def validate_layout(self) -> None:
        """Assert that ``base_dir`` exists and contains every layout directory.

        Raises:
            LayoutViolationError: if the base dir or any canonical directory is
                missing or is not a directory.
        """
        if not self.base_dir.is_dir():
            raise LayoutViolationError(
                f"base dir is missing or not a directory: {self.base_dir}"
            )
        for name in LAYOUT_DIRS:
            d = self.base_dir / name
            if not d.is_dir():
                raise LayoutViolationError(
                    f"missing canonical layout directory: {name!r}"
                )
# #EXT-006-REQ-2 End
